### PR TITLE
RESPA-220 | Database query optimizations for resource and reservation API endpoints

### PR DESCRIPTION
**Pull request by tommimanbytes**
Various query optimizations. Most of them are about preventing additional db queries by not using querysets `.filter()` or `.exists()` methods in places where they could raise query count cumulatively  in iterations. Also prefetching user models unit permissions.

**This optimization optimizes list endpoint queries of reservation and resource from this:**
Selection_071

**To this:**
Selection_072

### CODE REVIEW

File: resources/auth.py [hunk 1]
```diff
@@ -13,7 +13,36 @@ def is_general_admin(user):
 
 
 def is_any_admin(user):
-    return is_authenticated_user(user) and (
-        is_general_admin(user) or
-        user.unit_group_authorizations.admin_level().exists() or
-        user.unit_authorizations.admin_level().exists())
+    from .enums import UnitGroupAuthorizationLevel, UnitAuthorizationLevel
```
Comment (otiala) [inline resources/auth.py]: Is there a specific reason for the imports at function level?

Comment (tommimanbytes) [reply resources/auth.py]: Not really. I changed them to be at module level.

**Comment by jussiarpalahti** [comment]
Apropos, where do you got the pictures with timings and query counts, @tommimanbytes ?
 Looks very useful.

**Comment by tommimanbytes** [comment]
> Apropos, where do you got the pictures with timings and query counts, @tommimanbytes ?
> Looks very useful.

I used Django Silk library. It's a library that records all requests made to API and provides a web UI with detailed info about database queries made during the request.

[Django Silk Github page](https://github.com/jazzband/django-silk)

**Pull request merged**